base/utf16.jl

```diff
@@ -1,10 +1,82 @@
 # This file is a part of Julia. License is MIT: http://julialang.org/license
 
+#=
+@doc """
+@brief      Error messages for Unicode / UTF support
+""" ->
+=#
+#=
+@enum(UTF_ERR,
+      SHORT = 1,
+      CONT,
+      LONG,
+      NOT_LEAD,
+      NOT_TRAIL,
+      NOT_SURROGATE,
+      MISSING_SURROGATE,
+      INVALID,
+      SURROGATE,
+      NULL_TERMINATE,
+)
+=#
+
+const UTF_ERR_SHORT = 1
+const UTF_ERR_CONT  = 2
+const UTF_ERR_LONG  = 3
+const UTF_ERR_NOT_LEAD = 4
+const UTF_ERR_NOT_TRAIL = 5
+const UTF_ERR_NOT_SURROGATE = 6
+const UTF_ERR_MISSING_SURROGATE = 7
+const UTF_ERR_INVALID = 8
+const UTF_ERR_SURROGATE = 9
+const UTF_ERR_NULL_TERMINATE = 10
+const UTF_ERR_MAX = 10
+
+type UnicodeError <: Exception
+    msg::AbstractString
+end
+
+#=
+@doc """
+@brief      Throws UnicodeError with information about the specific error, location, and character
+
+@param[in]  errCode::UTF_ERR
+@param[in]  errPos:: Integer
+@param[in]  errChar::Integer
+
+@throws never returns, always throws UnicodeError
+""" ->
+=#
+function utf_errfunc(errCode::Integer, errPos::Integer, errChar::Integer)
+    const errMsgs = [
```

Comment:
Why not just throw the error at the callsite (use can then use the interpolated values the error message)?  You are not gaining much it seems by factoring this out into a separate function because most the error codes do not overlap at the call sites.  A bit more verbose but easier to understand IMO.